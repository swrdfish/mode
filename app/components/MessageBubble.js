import React from 'react'
import AutoLinker from 'autolinker'
import escapeHTML from 'escape-html'


class MessageBubble extends React.Component {
    constructor(props) {
        super(props)
    }

    render() {
        let item = this.props.item
        let className = item.isMine?'message-bubble mine':'message-bubble'
        let downloadButtonClass = item.isMine?"":"fa fa-download"
        let pretext = item.isMine?"Sent ":"Download "
        let text

        let itemRendered
        if(item.type == "file") {
            if(item.body.count > 1) {
                text = (
                    <a href="#" onClick={this.props.onDownload} data-id={item.body.id}>
                        {item.body.count} files&nbsp;&nbsp;&nbsp;
                        <span className={downloadButtonClass + " download-button"}/>
                    </a>
                )
            } else {
                text = (
                    <a href="#" onClick={this.props.onDownload} data-id={item.body.id}>
                        {item.body.filenames[0]}&nbsp;&nbsp;&nbsp;
                        <span className={downloadButtonClass + " download-button"}/>
                    </a>
                )
            }

            if(item.isMine) {
                text = "" + item.body.count
                text += item.body.count > 1?" files":" file"
            }
            className += " file"
            itemRendered = (
                <div className='message-bubble-wrapper'>
                    <div className={className}>
                        <span>
                            {pretext}&nbsp;&nbsp;{text}               
                        </span>
                        <span className='time-stamp'>{item.timeStamp.toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', hour12: true })}</span>
                    </div>
                </div>
            )
        } else {
            let messageBody = AutoLinker.link(escapeHTML(item.body), {className: "message-bubble-link"})
            itemRendered = (
                <div className='message-bubble-wrapper'>
                    <div className={className}>
                        <span className='message-bubble-body' dangerouslySetInnerHTML={{ __html: messageBody }}></span>
                        <span className='time-stamp'>{item.timeStamp.toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', hour12: true })}</span>
                    </div>
                </div>
            )
        }
        return itemRendered
    }
}

export default MessageBubble