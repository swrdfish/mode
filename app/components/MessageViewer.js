import React from 'react'
import MessageBubble from "./MessageBubble"


class MessageViewer extends React.Component {
    render(){
        if(this.props.chatter) {
            let renderedMsg = (<div className='empty-message-view'>Say something!</div>)
            if (this.props.messages) {
                let messages = this.props.messages.reduceRight((arr, item) => {
                    return arr.concat((<MessageBubble item={item} key={item.id} onDownload={this.props.onDownload}/>))
                }, [])
                renderedMsg = (
                    <div className='messages-scrollable'>
                        {messages}
                    </div>
                )
            }

            return (
                <div className='message-view-wrapper'>
                    {renderedMsg}
                </div>
            )
        } else {
            return (
                <div className='message-view-wrapper'>
                    <div className='empty-message-view'>select someone to chat with.</div>
                </div>
            )
        }
    }
}

export default MessageViewer