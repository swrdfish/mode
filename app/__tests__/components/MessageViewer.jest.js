import React from 'react'
import { shallow } from 'enzyme'
import MessageViewer from '../../components/MessageViewer'
import MessageBubble from '../../components/MessageBubble'
import renderer from 'react-test-renderer';


describe('Component: MessageViewer', () => {
    it('should render the message viewer', () => {
        const wrapper = shallow(<MessageViewer />)
        expect(wrapper.find(".message-view-wrapper")).toHaveLength(1)
        expect(wrapper.find(".empty-message-view")).toHaveLength(1)
        expect(wrapper.find(".empty-message-view").text()).toEqual("select someone to chat with.")
    })

    it('should match previous snapshot', () => {
        let chatter = 123456
        const tree = renderer
            .create(<MessageViewer chatter={chatter} />).toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('should render empty chat room if no messages are present', () => {
        let chatter = 123456
        const wrapper = shallow(<MessageViewer chatter={chatter} />)
        expect(wrapper.find(".message-view-wrapper")).toHaveLength(1)
        expect(wrapper.find(".empty-message-view")).toHaveLength(1)
        expect(wrapper.find(".empty-message-view").text()).toEqual("Say something!")  
    })

    it('should render messages', () => {
        let chatter = 123456

        let item1 = {
            type: "text",
            body: "hello world",
            id: 123,
            isMine: false,
            timeStamp: Date.now()
        }

        let item = {
            type: "text",
            body: "hello world",
            id: 124,
            isMine: true,
            timeStamp: Date.now()
        }

        let messages = [item, item1]
        const wrapper = shallow(<MessageViewer chatter={ chatter } messages={ messages } />)
        expect(wrapper.find(".message-view-wrapper")).toHaveLength(1)
        expect(wrapper.find(".empty-message-view")).toHaveLength(0)
        expect(wrapper.find(".messages-scrollable")).toHaveLength(1)
        expect(wrapper.find(MessageBubble)).toHaveLength(2)
    })
})