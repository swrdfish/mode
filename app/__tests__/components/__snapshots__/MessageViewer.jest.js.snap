// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: MessageViewer should match previous snapshot 1`] = `
<div
  className="message-view-wrapper"
>
  <div
    className="empty-message-view"
  >
    Say something!
  </div>
</div>
`;
